### PR TITLE
[NavigationDrawer] Add dynamically sizing support for the drawer + performance improvements

### LINKED ISSUE #5514: [NavigationDrawer] Allow content to dynamically change size
This was filed as an internal issue. If you are a Googler, please visit [b/118367614](http://b/118367614) for more details.

---
#### Internal data
- Associated internal bug: [b/118367614](http://b/118367614)

### PR DESCRIPTION
**Pull request by yarneo**
**Context:**
Clients need the ability to dynamically size their navigation drawer content and have the drawer work appropriately and correctly.

**The Problem:**
Currently we have no mechanism where we let the drawer know that the size of the content has changed and the internal calculations need to be re-calculated and re-applied.

Also, before we were partially listening to the `preferredContentSize` by asking if it has changed every so often using the conditional in the `contentHeightSurplus` getter. Some clients can have a very calculations heavy `preferredContentSize`, which will cause the whole drawer mechanism to lag.

**The Fix:**
Because the `contentViewController` and the `headerViewController` are child view controllers of `MDCBottomDrawerContainerViewController` , we can listen to `preferredContentSizeDidChangeForChildContentContainer` and reset the previously calculated variables and relayout the drawer.

**Testing:**
Unit Test + Tested on an iPhone X and iPhone 7 on all examples, portrait and landscape. A dragons example has been created exactly for the testing of the dynamically changing content size here: #5545 

**Related Bugs:**
Closes #5514

### CODE REVIEW

File: components/NavigationDrawer/src/private/MDCBottomDrawerContainerViewController.m [hunk 1]
```diff
@@ -439,6 +439,14 @@ - (void)viewDidDisappear:(BOOL)animated {
   self.headerShadowLayer = nil;
 }
 
+- (void)preferredContentSizeDidChangeForChildContentContainer:(id<UIContentContainer>)container {
+  [super preferredContentSizeDidChangeForChildContentContainer:container];
+  _contentHeaderTopInset = NSNotFound;
+  _contentHeightSurplus = NSNotFound;
+  _addedContentHeight = NSNotFound;
```
Comment (jverkoey) [inline components/NavigationDrawer/src/private/MDCBottomDrawerContainerViewController.m:446]: This is a peculiar use of NSNotFound. Did you consider using a bool to indicate whether or not these values are valid instead?

NSNotFound docs:
> A value indicating that a requested item couldn’t be found or doesn’t exist.
>
> NSNotFound is typically used by various methods and functions that search for items in serial data and return indices, such as characters in a string object or id objects in an NSArray object.

File: components/NavigationDrawer/tests/unit/MDCNavigationDrawerScrollViewTests.m [hunk 2]
```diff
@@ -463,4 +463,19 @@ - (void)testBottomDrawerCornersAPIFullScreen {
...
+      [[MDCNavigationDrawerFakeTableViewController alloc] init];
+  [self.fakeBottomDrawer viewDidLoad];
+  [self.fakeBottomDrawer cacheLayoutCalculations];
+  CGFloat previousContentHeaderTopInset = self.fakeBottomDrawer.contentHeaderTopInset;
+
+  // When
+  self.fakeBottomDrawer.contentViewController.preferredContentSize = CGSizeMake(200, 200);
+
+  // Then
+  XCTAssertNotEqual(previousContentHeaderTopInset, self.fakeBottomDrawer.contentHeaderTopInset);
```
Comment (jverkoey) [inline components/NavigationDrawer/tests/unit/MDCNavigationDrawerScrollViewTests.m]: To narrow down this test a bit, can we verify that it's greater than or less than the previous value?

File: components/NavigationDrawer/src/private/MDCBottomDrawerContainerViewController.m (hunk 1 above)
Comment (yarneo) [reply components/NavigationDrawer/src/private/MDCBottomDrawerContainerViewController.m:446]: I agree that this use isn't optimal, but to change that mechanism I'll have to do a big cleanup as the whole code relies on NSNotFound for its properties. This is outside the scope of this PR. I've opened https://github.com/material-components/material-components-ios/issues/5635 for tracking.

File: components/NavigationDrawer/tests/unit/MDCNavigationDrawerScrollViewTests.m (hunk 2 above)
Comment (yarneo) [reply components/NavigationDrawer/tests/unit/MDCNavigationDrawerScrollViewTests.m]: done

**Review by andrewoverton — APPROVED** [review]
Conflicts :(

**Pull request merged**